### PR TITLE
Add reencrypt job to upgrade route

### PR DESCRIPTION
**Pull request by paulbarfuss**
Hey @guimou  I am putting this up as a draft for discussion, etc..

It is close to something that could work. I have been unable to fully test it in my own environment but can tell you the job runs right away before the `service` certs are created.

It could be a Cronjob running every 10 minutes or similar but ideally handled down the road by an operator with watches on the route and tls secrets.

### CODE REVIEW

**Comment by guimou** [comment]
The certs are generated through an init container when the StatefulSet is deployed. I don't think it could be another initContainer as the Pod is not running yet, so there is no Service to call.
Could this work with just a sidecar, waiting for the Service to be up, which would then patch the Route and stop?
Or do we have post hooks in a StatefulSet, that would trigger when the Readiness flag is set?

**Comment by paulbarfuss** [comment]
Having this run a sidecar may be a step in the right direction. I realized after opening this PR that the `openssl` command only works on the route when it is in passthrough mode, so getting the certificates directly from the filesystem would be the way. Or possibly hitting the Service directly.

I could also look into storing the Service certificates in a `kubernetes.io/tls` Secret and then mount it to the StatefulSet? That would likely enable `cert-utils-operator` to keep the `DestinationCA` updated on the route with an annotation: `cert-utils-operator.redhat-cop.io/destinationCA-from-secret: "<secret-name>"`

**Comment by guimou** [comment]
> Having this run a sidecar may be a step in the right direction. I realized after opening this PR that the `openssl` command only works on the route when it is in passthrough mode, so getting the certificates directly from the filesystem would be the way. Or possibly hitting the Service directly.

I'd definitely hit the Service directly. No need to setup a Route (plus go through the Router) just to get the certificates.

> I could also look into storing the Service certificates in a `kubernetes.io/tls` Secret and then mount it to the StatefulSet? That would likely enable `cert-utils-operator` to keep the `DestinationCA` updated on the route with an annotation: `cert-utils-operator.redhat-cop.io/destinationCA-from-secret: "<secret-name>"`

That's what I looked into at the beginning. The problem with this method is that it implies making modifications on the NiFi part (retrieve the certs, merge them in the TrustStore/KeyStore,...), which prevents from using the "vanilla" container image that is distributed by the project. So more maintenance. I'd rather stay as close as possible to a standard NiFi deployment.

**Comment by paulbarfuss** [comment]
I am giving this some more thought and using an initContainer might work very well. If we ran something similar to the shell script in this PR just after the nifi toolkit, we should be able to just insert the contents of `/opt/nifi/nifi-current/config-data/certs/nifi-cert.pem` into the DestinationCA without needing to run the `openssl` command to fetch that from the service. This would be most likely the shortest path to success without involving an external operator.

It might still be an added bonus to drop the nifi-cert.pem into a secret for _potential_ use by cert-utils-operator for long-lived deployments.

**Comment by guimou** [comment]
You're right, of course, no need to wait for the service to start!
So can be either having another initContainer right after the toolkit, or
add some more steps in the existing one. The latter is cleaner, the former
allows not to touch the original script that comes from standard Nifi
deployment (so that people are not messed up).
But now that I think of it tbh, people have to make some efforts to deploy
this and understand what's going on anyway. And this deployment is already
tweaked for OpenShift. So I guess inserting the code directly inside the
toolkit container would make sense.

**Comment by paulbarfuss** [comment]
Cool cool, I have updated to:
- move reencrypt code into a new initContainer just after the certs are created
- point the ClusterRoleBinding to the `nifi-sa` service account
- remove the reencrypt job (also from kustomize resources)
Should be a big step in the right direction. I would imagine GitOps tools would try to overwrite to Route with the contents of the passthrough route, so still maybe not the perfect solution but better.

Also, for various reasons not having a great time testing this on my local cluster. Might it be best to attempt a second deployment in a new namespace to test?

File: nifi/base/reencrypt-rbac.yaml [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+metadata:
+  name: route-reencrypt-role
+rules:
+  - apiGroups:
+      - ""
+      - route.openshift.io
+    resources:
+      - routes
+      - routes/custom-host
+    verbs:
```
Comment (guimou) [inline nifi/base/reencrypt-rbac.yaml]: You can only keep:
- get
- list
- update
- patch
which will reduce the openings

File: nifi/base/reencrypt-rbac.yaml [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+      - create
+      - get
+      - list
+      - watch
+      - update
+      - patch
+      - delete
+  - apiGroups:
```
Comment (guimou) [inline nifi/base/reencrypt-rbac.yaml]: You can remove this section (see in other files why)

File: nifi/base/ss-nifi.yaml [hunk 3]
```diff
@@ -95,6 +95,60 @@ spec:
...
+        - |
+          while [ true ]; do 
+            oc get route/${ROUTE_NAME} -n ${ROUTE_NAMESPACE}
+            if [ $? -eq 0 ]; then
+              ROUTE_HOSTNAME=$(oc get route/${ROUTE_NAME} -n ${ROUTE_NAMESPACE} -o jsonpath="{.spec.host}")
+              break
+            fi 
+            sleep 5s
+          done
+          oc extract secret/${ROUTER_CERTS_SECRET} -n openshift-ingress --to=/tmp
```
Comment (guimou) [inline nifi/base/ss-nifi.yaml]: Secret is not needed

File: nifi/base/ss-nifi.yaml [hunk 4]
```diff
@@ -95,6 +95,60 @@ spec:
...
+            host: ${ROUTE_HOSTNAME}
+            port:
+              targetPort: https
+            tls:
+              termination: Reencrypt
+              insecureEdgeTerminationPolicy: Redirect
+              key: |-
+          $(sed 's/^/      /' /tmp/tls.key)
+              certificate: |-
+          $(sed 's/^/      /' /tmp/tls.crt)
```
Comment (guimou) [inline nifi/base/ss-nifi.yaml]: The key and certificate sections ended empty in the Route anyway. They are not needed, I guess the default Route certificate is applied, which is enough for us.

**Review by guimou — COMMENTED** [review]
I also modified my Route definition this way to make it clearer:
```
tls:
    termination: reencrypt
    insecureEdgeTerminationPolicy: Redirect
    destinationCACertificate: |-
      TO_BE_REPLACED
      
```

**Comment by paulbarfuss** [comment]
Agree 100% with all of the above, and would also like to dial that ClusterRole back down to a namespaced Role since we don't need to reach outside for secrets.

File: nifi/base/reencrypt-rbac.yaml (hunk 1 above)
Comment (paulbarfuss) [reply nifi/base/reencrypt-rbac.yaml]: I had made an assumption that replace to use either create or patch but seems that it has its own verb... replace. Going to clean up and resubmit.

**Comment by paulbarfuss** [comment]
I believe the RBAC is more-or-less good now:
```bash
❯ oc policy who-can replace routes -n nifi
resourceaccessreviewresponse.authorization.openshift.io/<unknown> 

Namespace: nifi
Verb:      replace
Resource:  routes.route.openshift.io

Users:  kubeadmin
        system:admin
        system:serviceaccount:nifi:nifi-sa
etc...
```
My assumption as noted in the comments was this would be one of either:
- create, or
- patch
Looks like it has its own verb. Do we want to try to move this into a patch? Either way I think we are close here.

**Pull request merged**